Custom RPC incorrectly inserts chainId(4-rinkeby) eth_signTypedData_v3, signing tx affected.

**Describe the bug**

**To Reproduce**
Steps to reproduce the behavior:
1. Set Custom RPC URL to `https://api.myetherwallet.com/eth `
2. Add some extra custom RPC such as `https://api.myetherwallet.com/rop`
3. Go to https://rstormsf.github.io/js-eth-personal-sign-examples/
4. Click Connect
5. Click sign typed data v3
6. See error
```
AssertionError: Provided chainId (1) must match the active chainId (4)
```

**Expected behavior**
Custom RPC must not use hardcoded values

**Browser details (please complete the following information):**
 - OS: [OSX]
 - Browser [chrome]
 - MetaMask Version [5.2.2]

**Additional context**

I have created PR to remove hardcoded chainId
https://github.com/danfinlay/js-eth-personal-sign-examples/pull/7

![image](https://user-images.githubusercontent.com/9360827/50310219-248f0f80-04c3-11e9-9f74-105caa35307c.png)


I also wasn't able to send any transactions from custom RPC URLs, and when I click on Tx Hash, it redirects me to rinkeby.etherscan.io 
I believe it's ignoring custom rpcs and reading from last network id from dropdown which is rinkeby

--- a/app/scripts/controllers/preferences.js
+++ b/app/scripts/controllers/preferences.js
@@ -406,7 +406,6 @@ class PreferencesController {
   updateRpc (newRpcDetails) {
     const rpcList = this.getFrequentRpcListDetail()
     const index = rpcList.findIndex((element) => { return element.rpcUrl === newRpcDetails.rpcUrl })
-
     if (index > -1) {
       const rpcDetail = rpcList[index]
       const updatedRpc = extend(rpcDetail, newRpcDetails)

--- a/app/scripts/controllers/preferences.js
+++ b/app/scripts/controllers/preferences.js
@@ -392,6 +392,33 @@ class PreferencesController {
   }
 
   /**
+   * updates custom RPC details
+   *
+   * @param {string} url The RPC url to add to frequentRpcList.
+   * @param {number} chainId Optional chainId of the selected network.
+   * @param {string} ticker   Optional ticker symbol of the selected network.
+   * @param {string} nickname Optional nickname of the selected network.
+   * @returns {Promise<array>} Promise resolving to updated frequentRpcList.
+   *
+   */
+
+
+  updateRpc (newRpcDetails) {
+    const rpcList = this.getFrequentRpcListDetail()
+    const index = rpcList.findIndex((element) => { return element.rpcUrl === newRpcDetails.rpcUrl })
+
+    if (index > -1) {
+      const rpcDetail = rpcList[index]
+      const updatedRpc = extend(rpcDetail, newRpcDetails)
+      rpcList[index] = updatedRpc
+      this.store.updateState({ frequentRpcListDetail: rpcList })
+    } else {
+      const { url, chainId, ticker, nickname } = newRpcDetails
+      this.addToFrequentRpcList(url, chainId, ticker, nickname)
+    }
+
+  }
+  /**
    * Adds custom RPC url to state.
    *
    * @param {string} url The RPC url to add to frequentRpcList.

--- a/app/scripts/controllers/preferences.js
+++ b/app/scripts/controllers/preferences.js
@@ -412,10 +412,10 @@ class PreferencesController {
       rpcList[index] = updatedRpc
       this.store.updateState({ frequentRpcListDetail: rpcList })
     } else {
-      const { url, chainId, ticker, nickname } = newRpcDetails
-      this.addToFrequentRpcList(url, chainId, ticker, nickname)
+      const { rpcUrl, chainId, ticker, nickname } = newRpcDetails
+      return this.addToFrequentRpcList(rpcUrl, chainId, ticker, nickname)
     }
-
+    return Promise.resolve(rpcList)
   }
   /**
    * Adds custom RPC url to state.

--- a/app/scripts/metamask-controller.js
+++ b/app/scripts/metamask-controller.js
@@ -414,6 +414,7 @@ module.exports = class MetamaskController extends EventEmitter {
       // network management
       setProviderType: nodeify(networkController.setProviderType, networkController),
       setCustomRpc: nodeify(this.setCustomRpc, this),
+      updateAndSetCustomRpc: nodeify(this.updateAndSetCustomRpc, this),
       delCustomRpc: nodeify(this.delCustomRpc, this),
 
       // PreferencesController
@@ -1501,6 +1502,21 @@ module.exports = class MetamaskController extends EventEmitter {
   }
 
   // network
+  /**
+   * A method for selecting a custom URL for an ethereum RPC provider and updating it
+   * @param {string} rpcUrl - A URL for a valid Ethereum RPC API.
+   * @param {number} chainId - The chainId of the selected network.
+   * @param {string} ticker - The ticker symbol of the selected network.
+   * @param {string} nickname - Optional nickname of the selected network.
+   * @returns {Promise<String>} - The RPC Target URL confirmed.
+   */
+
+  async updateAndSetCustomRpc (rpcUrl, chainId, ticker = 'ETH', nickname) {
+    await this.preferencesController.updateRpc({ rpcUrl, chainId, ticker, nickname })
+    this.networkController.setRpcTarget(rpcUrl, chainId, ticker, nickname)
+    return rpcUrl
+  }
+
 
   /**
    * A method for selecting a custom URL for an ethereum RPC provider.

--- a/app/scripts/metamask-controller.js
+++ b/app/scripts/metamask-controller.js
@@ -1511,8 +1511,15 @@ module.exports = class MetamaskController extends EventEmitter {
    * @returns {Promise<String>} - The RPC Target URL confirmed.
    */
   async setCustomRpc (rpcTarget, chainId, ticker = 'ETH', nickname = '') {
-    this.networkController.setRpcTarget(rpcTarget, chainId, ticker, nickname)
-    await this.preferencesController.addToFrequentRpcList(rpcTarget, chainId, ticker, nickname)
+    const frequentRpcListDetail = this.preferencesController.getFrequentRpcListDetail()
+    const rpcSettings = frequentRpcListDetail.find((rpc) => rpcTarget === rpc.rpcUrl)
+
+    if (rpcSettings) {
+      this.networkController.setRpcTarget(rpcSettings.rpcUrl, rpcSettings.chainId, rpcSettings.ticker, rpcSettings.nickname)
+    } else {
+      this.networkController.setRpcTarget(rpcTarget, chainId, ticker, nickname)
+      await this.preferencesController.addToFrequentRpcList(rpcTarget, chainId, ticker, nickname)
+    }
     return rpcTarget
   }
 

--- a/test/unit/app/controllers/preferences-controller-test.js
+++ b/test/unit/app/controllers/preferences-controller-test.js
@@ -515,6 +515,9 @@ describe('preferences controller', function () {
     it('should update the rpcDetails properly', () => {
       preferencesController.store.updateState({frequentRpcListDetail: [{}, { rpcUrl: 'test' }, {}]})
       preferencesController.updateRpc({ rpcUrl: 'test', chainId: 1 })
+      preferencesController.updateRpc({ rpcUrl: 'test/1', chainId: 1 })
+      preferencesController.updateRpc({ rpcUrl: 'test/2', chainId: 1 })
+      preferencesController.updateRpc({ rpcUrl: 'test/3', chainId: 1 })
       const list = preferencesController.getFrequentRpcListDetail()
       assert.deepEqual(list[1], { rpcUrl: 'test', chainId: 1 })
     })

--- a/test/unit/app/controllers/preferences-controller-test.js
+++ b/test/unit/app/controllers/preferences-controller-test.js
@@ -511,6 +511,15 @@ describe('preferences controller', function () {
     })
   })
 
+  describe('#updateRpc', function () {
+    it('should update the rpcDetails properly', () => {
+      preferencesController.store.updateState({frequentRpcListDetail: [{}, { rpcUrl: 'test' }, {}]})
+      preferencesController.updateRpc({ rpcUrl: 'test', chainId: 1 })
+      const list = preferencesController.getFrequentRpcListDetail()
+      assert.deepEqual(list[1], { rpcUrl: 'test', chainId: 1 })
+    })
+  })
+
   describe('on updateFrequentRpcList', function () {
     it('should add custom RPC url to state', function () {
       preferencesController.addToFrequentRpcList('rpc_url', 1)

--- a/ui/app/actions.js
+++ b/ui/app/actions.js
@@ -1893,7 +1893,7 @@ function setPreviousProvider (type) {
 function updateAndSetCustomRpc (newRpc, chainId, ticker = 'ETH', nickname) {
   return (dispatch) => {
     log.debug(`background.updateAndSetCustomRpc: ${newRpc} ${chainId} ${ticker} ${nickname}`)
-    background.updateAndSetCustomRpc(newRpc, chainId, ticker, nickname, (err, result) => {
+    background.updateAndSetCustomRpc(newRpc, chainId, ticker, nickname || newRpc, (err, result) => {
       if (err) {
         log.error(err)
         return dispatch(actions.displayWarning('Had a problem changing networks!'))
@@ -1906,10 +1906,10 @@ function updateAndSetCustomRpc (newRpc, chainId, ticker = 'ETH', nickname) {
   }
 }
 
-function setRpcTarget (newRpc, chainId, ticker = 'ETH', nickname = '') {
+function setRpcTarget (newRpc, chainId, ticker = 'ETH', nickname) {
   return (dispatch) => {
     log.debug(`background.setRpcTarget: ${newRpc} ${chainId} ${ticker} ${nickname}`)
-    background.setCustomRpc(newRpc, chainId, ticker, nickname, (err, result) => {
+    background.setCustomRpc(newRpc, chainId, ticker, nickname || newRpc, (err, result) => {
       if (err) {
         log.error(err)
         return dispatch(actions.displayWarning('Had a problem changing networks!'))

--- a/ui/app/actions.js
+++ b/ui/app/actions.js
@@ -237,6 +237,7 @@ var actions = {
   updateTokens,
   removeSuggestedTokens,
   UPDATE_TOKENS: 'UPDATE_TOKENS',
+  updateAndSetCustomRpc: updateAndSetCustomRpc,
   setRpcTarget: setRpcTarget,
   delRpcTarget: delRpcTarget,
   setProviderType: setProviderType,
@@ -1886,6 +1887,22 @@ function setPreviousProvider (type) {
   return {
     type: actions.SET_PREVIOUS_PROVIDER,
     value: type,
+  }
+}
+
+function updateAndSetCustomRpc (newRpc, chainId, ticker = 'ETH', nickname) {
+  return (dispatch) => {
+    log.debug(`background.updateAndSetCustomRpc: ${newRpc} ${chainId} ${ticker} ${nickname}`)
+    background.updateAndSetCustomRpc(newRpc, chainId, ticker, nickname, (err, result) => {
+      if (err) {
+        log.error(err)
+        return dispatch(actions.displayWarning('Had a problem changing networks!'))
+      }
+      dispatch({
+        type: actions.SET_RPC_TARGET,
+        value: newRpc,
+      })
+    })
   }
 }
 

--- a/ui/app/components/dropdowns/network-dropdown.js
+++ b/ui/app/components/dropdowns/network-dropdown.js
@@ -277,7 +277,6 @@ NetworkDropdown.prototype.getNetworkName = function () {
 NetworkDropdown.prototype.renderCommonRpc = function (rpcListDetail, provider) {
   const props = this.props
   const reversedRpcListDetail = rpcListDetail.slice().reverse()
-  const network = props.network
 
   return reversedRpcListDetail.map((entry) => {
     const rpc = entry.rpcUrl
@@ -288,7 +287,7 @@ NetworkDropdown.prototype.renderCommonRpc = function (rpcListDetail, provider) {
     if ((rpc === 'http://localhost:8545') || currentRpcTarget) {
       return null
     } else {
-      const chainId = entry.chainId || network
+      const chainId = entry.chainId
       return h(
         DropdownMenuItem,
         {

--- a/ui/app/components/pages/settings/settings-tab/settings-tab.component.js
+++ b/ui/app/components/pages/settings/settings-tab/settings-tab.component.js
@@ -230,7 +230,6 @@ export default class SettingsTab extends PureComponent {
 
   validateRpc (newRpc, chainId, ticker = 'ETH', nickname) {
     const { setRpcTarget, displayWarning } = this.props
-
     if (validUrl.isWebUri(newRpc)) {
       if (!!chainId && Number.isNaN(parseInt(chainId))) {
         return displayWarning(`${this.context.t('invalidInput')} chainId`)

--- a/ui/app/components/pages/settings/settings-tab/settings-tab.container.js
+++ b/ui/app/components/pages/settings/settings-tab/settings-tab.container.js
@@ -4,7 +4,7 @@ import { connect } from 'react-redux'
 import { withRouter } from 'react-router-dom'
 import {
   setCurrentCurrency,
-  setRpcTarget,
+  updateAndSetCustomRpc,
   displayWarning,
   revealSeedConfirmation,
   setUseBlockie,
@@ -50,7 +50,7 @@ const mapStateToProps = state => {
 const mapDispatchToProps = dispatch => {
   return {
     setCurrentCurrency: currency => dispatch(setCurrentCurrency(currency)),
-    setRpcTarget: (newRpc, chainId, ticker, nickname) => dispatch(setRpcTarget(newRpc, chainId, ticker, nickname)),
+    setRpcTarget: (newRpc, chainId, ticker, nickname) => dispatch(updateAndSetCustomRpc(newRpc, chainId, ticker, nickname)),
     displayWarning: warning => dispatch(displayWarning(warning)),
     revealSeedConfirmation: () => dispatch(revealSeedConfirmation()),
     setUseBlockie: value => dispatch(setUseBlockie(value)),

